torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -431,23 +437,24 @@ int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
   Variable sliced = applySlicing(
     self_, holder.get(), variableIndices, /*is_tracing=*/is_tracing, self_device, self_.sizes(), specified_dims);
   if (variableIndices.empty()) {
+    pybind11::gil_scoped_release no_gil;
     at::indexing::copy_to(sliced, value);
     return 0;
   }
 
-  IntArrayRef valueSizes = value.sizes();
-  IntArrayRef slicedValueSizes = at::indexing::slicePrefix1sSize(valueSizes);
-  torch::autograd::Variable valuesSliced;
-  if (!valueSizes.equals(slicedValueSizes)) {
-    valuesSliced = value.view(slicedValueSizes);
-  } else {
-    valuesSliced = value;
-  }
   {
     pybind11::gil_scoped_release no_gil;
+    IntArrayRef valueSizes = value.sizes();
+    IntArrayRef slicedValueSizes = at::indexing::slicePrefix1sSize(valueSizes);
+    torch::autograd::Variable valuesSliced;
+    if (!valueSizes.equals(slicedValueSizes)) {
+      valuesSliced = value.view(slicedValueSizes);
+    } else {
+      valuesSliced = value;
+    }
     at::indexing::dispatch_index_put_(sliced, std::move(variableIndices), valuesSliced);
+    return 0;
```

Comment:
There are other python-related objects in the scope so we do want to re-acquire the GIL before exiting no?
Or the order in which the destructors for local objects are being called guarantee us that this is ok?

If this needs changing, there is other occurrences above as well.

Reply:
Stack variables are always destroyed in reverse order of object creation so the gil will be re-acquired before any python-related calls.